embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/comms/delivery/EmbraceCacheService.kt

```diff
@@ -96,68 +116,111 @@ internal class EmbraceCacheService(
+        fileLocks.remove(name)
+        return success
     }
 
-    override fun listFilenamesByPrefix(prefix: String): List<String> {
+    override fun listFilenamesByPrefix(prefix: String): List<String> { // TODO: locking?
         return storageService.listFiles { _, name ->
             name.startsWith(EMBRACE_PREFIX + prefix)
         }.map { file -> file.name.substring(EMBRACE_PREFIX.length) }
     }
 
-    override fun writeSession(name: String, sessionMessage: SessionMessage) {
-        try {
-            logger.logDeveloper(TAG, "Attempting to write bytes to $name")
-            val file = storageService.getFileForWrite(EMBRACE_PREFIX + name)
-            serializer.toJson(sessionMessage, SessionMessage::class.java, file.outputStream())
-            logger.logDeveloper(TAG, "Bytes cached")
-        } catch (ex: Throwable) {
-            logger.logWarning("Failed to write session with buffered writer", ex)
+    override fun writeSession(name: String, sessionMessage: SessionMessage) = cacheObject(name, sessionMessage, SessionMessage::class.java)
+
+    override fun loadOldPendingApiCalls(name: String): List<PendingApiCall>? {
+        findLock(name).read {
+            val file = storageService.getFileForRead(EMBRACE_PREFIX + name)
+            try {
+                val type = Types.newParameterizedType(List::class.java, PendingApiCall::class.java)
+                return serializer.fromJson(file.inputStream(), type) as List<PendingApiCall>?
+                    ?: emptyList()
+            } catch (ex: FileNotFoundException) {
+                logger.logDebug("Cache file cannot be found " + file.path)
+            } catch (ex: Exception) {
+                logger.logDebug("Failed to read cache object " + file.path, ex)
+            }
+            return null
         }
     }
 
-    override fun loadOldPendingApiCalls(name: String): List<PendingApiCall>? {
-        val file = storageService.getFileForRead(EMBRACE_PREFIX + name)
-        try {
-            val type = Types.newParameterizedType(List::class.java, PendingApiCall::class.java)
-            return serializer.fromJson(file.inputStream(), type) as List<PendingApiCall>? ?: emptyList()
-        } catch (ex: FileNotFoundException) {
-            logger.logDebug("Cache file cannot be found " + file.path)
-        } catch (ex: Exception) {
-            logger.logDebug("Failed to read cache object " + file.path, ex)
+    override fun replaceSession(name: String, transformer: (SessionMessage) -> SessionMessage) {
+        findLock(name).write {
+            try {
+                val sessionMessage = loadObject(name, SessionMessage::class.java) ?: return@write
+                val sessionFile = storageService.getFileForWrite(EMBRACE_PREFIX + name)
+                val oldSessionFile = storageService.getFileForWrite("${sessionFile.name}-old")
+                val newSessionFile = storageService.getFileForWrite("${sessionFile.name}-new")
```

Comment:
Is there any risk that these get picked up by the cached session logic in `getAllCachedSessionIds`, or something similar? From what I can see, that just checks if a file contains a prefix & then assumes it's a session.

Reply:
I've rewritten the equivalent logic to `getAllCachedSessionIds` in the next PR such that we sweep for these temp files and deal with them, so we should never get into a case where the temps files are being treated as real session files